Zend/tests/magic_methods_inheritance_rules.phpt

```diff
@@ -0,0 +1,50 @@
+--TEST--
+Magic Methods inheritance rules
+--FILE--
+<?php
+class ValidMagicMethods {
+	public function __call(string $name, array $arguments): mixed {}
+
+	public static function __callStatic(string $name, array $arguments): mixed {}
+
+	public function __clone(): void {}
+
+	public function __debugInfo(): ?array {}
+
+	public function __get(string $name): mixed {}
+
+	public function __isset(string $name): bool {}
+
+	public function __serialize(): array {}
+
+	public function __set(string $name, mixed $value): void {}
+
+	public static function __set_state(array $properties): object {}
+
+	public function __sleep(): array {}
+
+	public function __unserialize(array $data): void {}
+
+	public function __unset(string $name): void {}
+
+	public function __wakeup(): void {}
+}
+
+final class NarrowedReturnType extends ValidMagicMethods {
+	public function __call(string $name, array $arguments): string|float|null {}
+
+	public static function __callStatic(string $name, array $arguments): ?array {}
+
+	public function __debugInfo(): array {}
+
+	public function __get(string $name): int|string {}
+}
+
+final class NarrowedArgumentType extends ValidMagicMethods {
+	// public function __get(string|array $name): mixed {}
```

Comment:
Union Types gurus: how can I make this thing possible? Looks like

```c
} else if (ZEND_TYPE_IS_SET(fptr->common.arg_info[0].type) && ZEND_TYPE_FULL_MASK(fptr->common.arg_info[0].type) != MAY_BE_STRING) {
```

isn't doing the trick :(

Reply:
To check whether `string` is a member of the union, you can query the status of the respective bit, like:
````.c
} else if (ZEND_TYPE_IS_SET(fptr->common.arg_info[0].type) && !(ZEND_TYPE_FULL_MASK(fptr->common.arg_info[0].type) & MAY_BE_STRING)) {
````

Reply:
@cmb69 @kocsismate It worked, thanks boys!